File: 1818-maximum-score-from-removing-substrings/1818-maximum-score-from-removing-substrings.java
class Solution {
    private int pairs;
    private StringBuilder carb;
    private int clear(String match, int pts) {
        int total_pts = 0;
        int w_idx = 0;

        for (int r_idx = 0; r_idx < carb.length(); r_idx++) {
            carb.setCharAt(w_idx++, carb.charAt(r_idx));

            if (w_idx < 2) continue;
            if (carb.charAt(w_idx - 2) == match.charAt(0) && carb.charAt(w_idx - 1) == match.charAt(1)) {
                w_idx -= 2;
                total_pts += pts;
            } 
        }
        
        carb.setLength(w_idx);

        return total_pts;
    }
    public int maximumGain(String s, int x, int y) {
        pairs = 0;
        carb = new StringBuilder(s);

        if (x > y) {
            pairs += clear("ab", x);
            pairs += clear("ba", y);
        } else {
            pairs += clear("ba", y);
            pairs += clear("ab", x);
        }

        return pairs;
    }
}

File: 1818-maximum-score-from-removing-substrings/1818-maximum-score-from-removing-substrings.cpp
class Solution {
private:
    int pairs;
    int clear(string& carb, string match, int pts) {
        int total_pts = 0;
        int w_idx = 0;

        for (int r_idx = 0; r_idx < carb.size(); r_idx++) {
            carb[w_idx++] = carb[r_idx];

            if (w_idx < 2) continue;
            if (carb[w_idx - 2] == match[0] && carb[w_idx - 1] == match[1]) {
                w_idx -= 2;
                total_pts += pts;
            } 
        }
        
        carb.erase(carb.begin() + w_idx, carb.end());

        return total_pts;
    }
public:
    int maximumGain(string s, int x, int y) {
        pairs = 0;

        if (x > y) {
            pairs += clear(s, "ab", x);
            pairs += clear(s, "ba", y);
        } else {
            pairs += clear(s, "ba", y);
            pairs += clear(s, "ab", x);
        }

        return pairs;
    }
};

auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();
